nx/source/services/vi.c

```diff
@@ -45,772 +57,315 @@ Result viInitialize(ViServiceType service_type)
-
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 1200;
-    raw->display_id = display->display_id;
-
-    Result rc = serviceIpcDispatch(&g_viISystemDisplayService);
-
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-            u64 z;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            if (z) *z = resp->z;
-        }
     }
 
-    return rc;
+    return serviceDispatchInOut(&g_viISystemDisplayService, 1200, display->display_id, *z);
 }
 
 Result viGetDisplayMaximumZ(ViDisplay *display, u64 *z) {
-    IpcCommand c;
-    ipcInitialize(&c);
-
-    if (!display->initialized)
+    if (!display->initialized) {
         return MAKERESULT(Module_Libnx, LibnxError_NotInitialized);
-
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        u64 display_id;
-    } *raw;
-
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 1202;
-    raw->display_id = display->display_id;
-
-    Result rc = serviceIpcDispatch(&g_viISystemDisplayService);
-
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-            u64 z;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            if (z) *z = resp->z;
-        }
     }
 
-    return rc;
+    return serviceDispatchInOut(&g_viISystemDisplayService, 1202, display->display_id, *z);
 }
 
-Result viCreateManagedLayer(const ViDisplay *display, ViLayerFlags layer_flags, u64 aruid, u64 *layer_id)
-{
-    IpcCommand c;
-    ipcInitialize(&c);
-
-    if (!display->initialized)
+Result viCreateManagedLayer(const ViDisplay *display, ViLayerFlags layer_flags, u64 aruid, u64 *layer_id) {
+    if (!display->initialized) {
         return MAKERESULT(Module_Libnx, LibnxError_NotInitialized);
+    }
 
-    struct {
-        u64 magic;
-        u64 cmd_id;
+    const struct {
         u32 layer_flags;
-        u32 pad;
         u64 display_id;
         u64 aruid;
-    } *raw;
-
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 2010;
-    raw->layer_flags = layer_flags;
-    raw->pad = 0;
-    raw->display_id = display->display_id;
-    raw->aruid = aruid;
-
-    Result rc = serviceIpcDispatch(&g_viIManagerDisplayService);
-
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-            u64 layer_id;
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            *layer_id = resp->layer_id;
-        }
-    }
-
-    return rc;
+    } in = { layer_flags, display->display_id, aruid };
+    return serviceDispatchInOut(&g_viIManagerDisplayService, 2010, in, *layer_id);
```

Comment:
This command has an aruid. I suspect `.in_send_pid = true,` is missing, but it was also missing in the old IPC version :thinking: